builtin/credential/aws-iam/cli.go

```diff
@@ -0,0 +1,137 @@
+package awsiam
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/vault/api"
+)
+
+type CLIHandler struct{}
+
+func transformHeaders(input map[string][]string) map[string]string {
+	retval := map[string]string{}
+	for k, v := range input {
+		retval[k] = v[0]
+	}
+	return retval
+}
+
+func (h *CLIHandler) Auth(c *api.Client, m map[string]string) (string, error) {
+	mount, ok := m["mount"]
+	if !ok {
+		mount = "aws-iam"
+	}
+
+	role, ok := m["role"]
+	if !ok {
+		role = ""
+	}
+
+	headerValue, ok := m["header_value"]
+	if !ok {
+		headerValue = ""
+	}
+
+	var stsSession *session.Session
+	var err error
+
+	access_key, ok := m["aws_access_key_id"]
+	if ok {
+		secret_key, ok := m["aws_secret_access_key"]
+		if !ok {
+			return "", fmt.Errorf("Explicitly specified aws_access_key_id but not aws_secret_access_key. Please specify them both.")
+		}
+		token, ok := m["aws_security_token"]
+		if !ok {
+			token = ""
+		}
+		staticCreds := credentials.NewStaticCredentials(access_key, secret_key, token)
+		stsSession, err = session.NewSessionWithOptions(session.Options{
+			Config: aws.Config{Credentials: staticCreds},
+		})
+		if err != nil {
+			return "", err
+		}
+	} else {
+		stsSession, err = session.NewSession()
+		if err != nil {
+			return "", err
+		}
+	}
+	var params *sts.GetCallerIdentityInput
+	svc := sts.New(stsSession)
+	stsRequest, _ := svc.GetCallerIdentityRequest(params)
+	if headerValue != "" {
+		stsRequest.HTTPRequest.Header.Add("X-Vault-Server", headerValue)
+	}
+	stsRequest.Sign()
+	headersJson, err := json.Marshal(transformHeaders(stsRequest.HTTPRequest.Header))
```

Comment:
No -- it's what you thought which is why you're not just using the built-in Header representation. I'm curious what Python does there if you have the same header multiple times -- does it just combine them into a single value with separated values?